File: Codes from the AI-assistants/[Experimental] mandelbrot/Basic Prompt - TBR Mode/Mandelbrot-gpt-basic-tbr.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

/* Usage:
   gcc -O3 -std=c11 -o mandelbrot_pbm mandelbrot_pbm.c
   ./mandelbrot_pbm out.pbm N max_iter xmin xmax ymin ymax
   Example:
   ./mandelbrot_pbm mandelbrot.pbm 1024 1000 -2.0 1.0 -1.5 1.5
*/

int main(int argc, char *argv[]) {
    if (argc < 4) {
        fprintf(stderr, "Usage: %s out.pbm N max_iter [xmin xmax ymin ymax]\n", argv[0]);
        return 1;
    }
    const char *outname = argv[1];
    int N = atoi(argv[2]);
    int max_iter = atoi(argv[3]);

    double xmin = -2.0, xmax = 1.0, ymin = -1.5, ymax = 1.5;
    if (argc >= 8) {
        xmin = atof(argv[4]);
        xmax = atof(argv[5]);
        ymin = atof(argv[6]);
        ymax = atof(argv[7]);
    }

    FILE *f = fopen(outname, "wb");
    if (!f) { perror("fopen"); return 2; }

    // PBM binary header (P4)
    fprintf(f, "P4\n%d %d\n", N, N);

    int bytes_per_row = (N + 7) / 8;
    unsigned char *rowbuf = malloc(bytes_per_row);
    if (!rowbuf) { perror("malloc"); fclose(f); return 3; }

    for (int py = 0; py < N; ++py) {
        // prepare empty buffer
        for (int b = 0; b < bytes_per_row; ++b) rowbuf[b] = 0;

        double ci = ymin + (double)py * (ymax - ymin) / (N - 1);
        for (int px = 0; px < N; ++px) {
            double cr = xmin + (double)px * (xmax - xmin) / (N - 1);

            double zr = 0.0, zi = 0.0;
            int iter = 0;
            while (iter < max_iter) {
                double zr2 = zr*zr;
                double zi2 = zi*zi;
                if (zr2 + zi2 > 4.0) break;
                double tmp = zr2 - zi2 + cr;
                zi = 2.0*zr*zi + ci;
                zr = tmp;
                ++iter;
            }

            int bit = (iter >= max_iter) ? 1 : 0; // 1 = black = in set
            if (bit) {
                int byte_idx = px / 8;
                int bit_idx = 7 - (px % 8); // PBM: MSB first
                rowbuf[byte_idx] |= (1 << bit_idx);
            }
        }
        // write row
        fwrite(rowbuf, 1, bytes_per_row, f);
    }

    free(rowbuf);
    fclose(f);
    return 0;
}

File: Codes from the AI-assistants/[Experimental] mandelbrot/Basic Prompt - TBR Mode/Mandelbrot-gpt-basic-tbr.py
#!/usr/bin/env python3
# Save as mandelbrot_pbm.py
# Usage:
#   python3 mandelbrot_pbm.py out.pbm N max_iter [xmin xmax ymin ymax]
# Example:
#   python3 mandelbrot_pbm.py mandelbrot.pbm 1024 1000 -2.0 1.0 -1.5 1.5

import sys
import math

def main():
    if len(sys.argv) < 4:
        print("Usage: {} out.pbm N max_iter [xmin xmax ymin ymax]".format(sys.argv[0]), file=sys.stderr)
        return 1

    outname = sys.argv[1]
    N = int(sys.argv[2])
    max_iter = int(sys.argv[3])

    if len(sys.argv) >= 8:
        xmin = float(sys.argv[4]); xmax = float(sys.argv[5])
        ymin = float(sys.argv[6]); ymax = float(sys.argv[7])
    else:
        xmin, xmax, ymin, ymax = -2.0, 1.0, -1.5, 1.5

    with open(outname, "wb") as f:
        f.write(b"P4\n%d %d\n" % (N, N))
        bytes_per_row = (N + 7) // 8

        for py in range(N):
            row = bytearray(bytes_per_row)
            ci = ymin + py * (ymax - ymin) / (N - 1)
            for px in range(N):
                cr = xmin + px * (xmax - xmin) / (N - 1)
                zr = 0.0
                zi = 0.0
                iter = 0
                while iter < max_iter:
                    zr2 = zr*zr
                    zi2 = zi*zi
                    if zr2 + zi2 > 4.0:
                        break
                    tmp = zr2 - zi2 + cr
                    zi = 2.0*zr*zi + ci
                    zr = tmp
                    iter += 1
                if iter >= max_iter: # inside -> black
                    bidx = px // 8
                    bit = 7 - (px % 8)
                    row[bidx] |= (1 << bit)
            f.write(row)

    return 0

if __name__ == "__main__":
    sys.exit(main())
